source/common/formatter/substitution_formatter.cc

```diff
@@ -54,5 +54,97 @@ const std::regex& SubstitutionFormatParser::commandWithArgsRegex() {
   // clang-format on
 }
 
+JsonFormatBuilder::FormatterEelements
+JsonFormatBuilder::fromStruct(const ProtobufWkt::Struct& struct_format) const {
+  sanitize_buffer_.clear();
+  raw_json_string_.clear();
+  output_.clear();
+
+  formatValueToFormatterEelements(struct_format);
+  if (!raw_json_string_.empty()) {
+    output_.push_back(JsonString{std::move(raw_json_string_)});
+  }
+  return std::move(output_);
+};
+
+void JsonFormatBuilder::formatValueToFormatterEelements(const ProtobufWkt::Value& value) const {
+  switch (value.kind_case()) {
+  case ProtobufWkt::Value::KIND_NOT_SET:
+  case ProtobufWkt::Value::kNullValue:
+    raw_json_string_.append(NullValue);
+    break;
+  case ProtobufWkt::Value::kNumberValue:
+    if (keep_value_type_) {
+      // Keep the number type in JSON.
+      raw_json_string_.append(fmt::to_string(value.number_value()));
+    } else {
+      // Convert the number to string.
+      raw_json_string_.append(fmt::format(R"("{}")", value.number_value()));
+    }
+    break;
+  case ProtobufWkt::Value::kStringValue: {
+    absl::string_view string_format = value.string_value();
+    if (string_format.empty() || !absl::StrContains(string_format, '%')) {
+      raw_json_string_.append(
+          fmt::format(R"("{}")", Json::sanitize(sanitize_buffer_, string_format)));
+    } else {
+      // The string contains a formatter, we need to push the current raw string
+      // into the output list first.
+      if (!raw_json_string_.empty()) {
+        output_.push_back(JsonString{std::move(raw_json_string_)});
+      }
+
+      // Now a formatter is coming, we need to push the current raw string into
+      // the output list.
+      output_.push_back(TmplString{std::string(string_format)});
+    }
+    break;
+  }
+  case ProtobufWkt::Value::kBoolValue:
+    if (keep_value_type_) {
+      // Keep the bool type in JSON.
+      raw_json_string_.append(fmt::to_string(value.bool_value()));
+    } else {
+      // Convert the bool to string.
+      raw_json_string_.append(fmt::format(R"("{}")", value.bool_value()));
+    }
+    break;
+  case ProtobufWkt::Value::kStructValue: {
+    formatValueToFormatterEelements(value.struct_value());
+    break;
+  case ProtobufWkt::Value::kListValue:
+    raw_json_string_.push_back('[');
+    for (int i = 0; i < value.list_value().values_size(); ++i) {
+      if (i > 0) {
+        raw_json_string_.push_back(',');
+      }
+      formatValueToFormatterEelements(value.list_value().values(i));
+    }
+    raw_json_string_.push_back(']');
+    break;
+  }
+  }
+}
+
+void JsonFormatBuilder::formatValueToFormatterEelements(const ProtobufWkt::Struct& value) const {
+  const auto& fields = value.fields();
+
+  // Sort the keys to make the output deterministic.
+  std::set<absl::string_view> keys;
+  std::for_each(fields.begin(), fields.end(), [&keys](const auto& p) { keys.insert(p.first); });
+
+  raw_json_string_.push_back('{');
+  for (auto it = keys.begin(); it != keys.end(); ++it) {
```

Comment:
WDYT of making a vector of name:value strings, sorting that, and combining it with absl::StrJoin. I think that might be cleaner and also faster due to absl::StrJoin about creating the correct sized buffer rather than checking for sizea for every token added to it.

Reply:
I think to use the vector is a good idea. But we couldn't use `absl::StrJoin` here. The value for map/vector entry here is substitution formatter that will be evaluated at runtime, not the final json value.

Reply:
I was thinking you'd make a vector of std::string and then join them.

It's a little subtle, but every time you append to raw_json_string you have to run conditions that check if it's big enough and resize it if it's not. Reducing the conditionals is the win here; with StrJoin, the absl library will compute the right size for the input and make the string big enough for that. Afterward it can just call memcpy on each substring.

Of course you still have to pay for the construction of each substring, which you are doing here in either case, so that is a wash.

Reply:
The value here may not be a raw string value. When loading configuration, the function here is used to construct `FormatterElements`, an array of raw JSON pieces and format string (string with substitution commands).

The final output is an array like this:

```
- '{"text":"text","tmpl":'                                       # Element 0, raw JSON piece.
- '%START_TIME% %FLAG%'                                          # Element 1, formatter string, must be independent element.
- ',"number":2,"bool":true,"list":["list_raw_value",false,'      # Element 2, raw JSON piece.
- '%EMIT_TIME%'                                                  # Element 3, formatter string, must be independent element.
- '],"nested":{"text":"nested_text"}}'                           # Element 4, raw JSON piece.
```

The keys, raw values, delimieters will be pushed to the raw_json_string_ first until we encounter a format string. Then the raw_json_string_ will be pushed into output array with type `JsonString`, and the format string will be pushed into output array with type `TmplString`.

So, I cannot image how can we use absl::StrJoin in this scenario.